src/asynch.rs

```diff
@@ -125,61 +149,88 @@ where
-                        ServerRecord::Handshake(ServerHandshake::NewSessionTicket(_)) => {
-                            // Ignore
-                            Ok(())
-                        }
-                        _ => {
-                            unimplemented!()
-                        }
-                    }?;
-                }
+                let record_data = if consumed < self.decrypted_len {
+                    // The current record is not completely consumed
+                    CryptoBuffer::wrap(
+                        &mut self.record_read_buf
+                            [self.decrypted_offset..self.decrypted_offset + self.decrypted_len],
+                    )
+                } else {
+                    // The current record is completely consumed, read the next...
+                    self.read_application_data().await?
+                };
+
+                let unread = &record_data.as_slice()[consumed..];
+                let to_copy = core::cmp::min(unread.len(), buf.len());
+                trace!("Got {} bytes to copy", to_copy);
+                buf[..to_copy].copy_from_slice(&unread[..to_copy]);
+                consumed += to_copy;
+                remaining -= to_copy;
             }
+            self.decrypted_consumed = consumed;
             Ok(buf.len() - remaining)
         } else {
             Err(TlsError::MissingHandshake)
         }
     }
 
+    async fn read_application_data<'m>(&'m mut self) -> Result<CryptoBuffer<'m>, TlsError> {
+        let socket = &mut self.delegate;
+        let key_schedule = &mut self.key_schedule;
+
+        let buf_ptr = self.record_read_buf.as_ptr();
+        let buf_len = self.record_read_buf.len();
+        let record =
+            decode_record::<Socket, CipherSuite>(socket, self.record_read_buf, key_schedule)
+                .await?;
+        let mut records = Queue::new();
+        decrypt_record::<CipherSuite>(key_schedule, &mut records, record)?;
+
+        while let Some(record) = records.dequeue() {
+            match record {
+                ServerRecord::ApplicationData(data) => {
+                    trace!("Got application data record");
+
+                    // SAFETY: Assyme `decrypt_record()` to decrypt in-place
+                    let slice = data.data.as_slice();
+                    let slice_ptr = slice.as_ptr();
+                    let offset = unsafe { slice_ptr.offset_from(buf_ptr) };
+                    assert!(offset >= 0);
+                    let offset = offset as usize;
+                    assert!(offset + slice.len() <= buf_len);
+
+                    self.decrypted_offset = offset;
+                    self.decrypted_len = slice.len();
+                    self.decrypted_consumed = 0;
+                    return Ok(data.data);
+                }
+                ServerRecord::Alert(alert) => {
+                    if let AlertDescription::CloseNotify = alert.description {
+                        Err(TlsError::ConnectionClosed)
```

Comment:
Re: our discussion earlier, this should probably set self.opened = false;

Reply:
Done